### ISSUE TITLE
POST v2/entities create a not-well formed context entity if attributes begin with double underline

### ISSUE THREAD

**Issue by LeonanCarvalho** [telefonicaid/fiware-orion#2453]
I didn't find any reference about that, I think it's a bug in entity creation.

```
POST http://localhost:1026/v2/entities
{
    "id": "KMZ1211",
    "type": "Bus",
    "__owner": {
        "value": 1,
        "type": "Integer",
        "metadata": {
            "client_id": {
                "type": "Integer",
                "value": "1cb7246a5fcd11e69cd1ecf4bbf955f4"
            }
        }
    },
    "__UID": {
        "value": 17,
        "type": "Integer",
        "metadata": {
            "class_uid": {
                "type": "Integer",
                "value": 1
            }
        }
    },
    "__label": {
        "value": "test",
        "type": "String"
    }
}
```

Result:

```
GET http://localhost:1026/v2/entities/KMZ1211
{
  "id": "KMZ1211",
  "type": "Bus",
  "": {
    "type": "Integer",
    "value": 1,
    "metadata": {
      "ID": {
        "type": "string",
        "value": "owner"
      },
      "client_id": {
        "type": "Integer",
        "value": "1cb7246a5fcd11e69cd1ecf4bbf955f4"
      }
    }
  }
}
```

OCB Info:
```
{
  "orion" : {
  "version" : "1.2.0-next",
  "uptime" : "9 d, 17 h, 27 m, 37 s",
  "git_hash" : "c1a224db08086751f05d471a8b52c0eaec15f3a5",
  "compile_time" : "Thu Aug 4 09:04:12 BRT 2016",
  "compiled_by" : "root"
}
}
```

**Comment by fgalan** [comment]
This is weird... Provisionaly labeles as bug, while we get some time to have a look to it.

Thanks for the feedback!

**Comment by kzangeli** [comment]
It seems the broker uses double underscore as an internal delimiter between 'attribute name' and metadata ID, probably for performance reasons.

From what I've seen, in mongo, the attribute name is stored as (only for attributes that have a metadata ID) as {attr-name}__{meta-ID}, and when finding an attribute with a name containing '__', the attribute name is taken from the LHS of '__' and the metadata ID is the RHS.

Now, the metadata ID ... does the broker still support that 'strange' construct?

If 'Yes', we'll need to implement a more complex inclusion of the metadata ID in mongo.
If 'No', let's just remove the piece of code that splits the name of the attribute between attr.name and meta-ID.

See the function `updateAttribute` for the creation of an attribute:

      /* Attributes with metadata ID are stored as <attrName>__<ID> in the attributes embedded document */
    std::string effectiveName = dbDotEncode(caP->name);
    if (caP->getId() != "")
    {
      effectiveName += "__" + caP->getId();
    }

The case of this issue is that the broker itself doesn't add the '__' between attr-name and meta-ID, but the very attrName from the client payload already has it.

Somewhere (I haven't looked it up) the broker detects the '__' and splits the string into attr-name (empty) and meta-ID (whatever follows the double underscore). 

Two choices:
- Forbid attr-names to contain a double underscore - return error
- Change the internal mechanism, to not use double underscore but something more 'ugly'

**Comment by fgalan** [comment]
It clearly seem that we should go for option 1. However, the decission should be aligned with all the other componentes in IoTP (as forbidden charset is a global aspect).

**Comment by gtorodelvalle** [comment]
I have been told (although I was not able to confirm it) that many Hydra attributes have this pattern, I mean, starting with 2 or more `_` characters. Just in case this influences the decision... :)

**Comment by fgalan** [comment]
Fixed in PR #2616 (release/1.4.0) and PR #2619 (master).
